### PR TITLE
Tiny wrapper for requests

### PR DESCRIPTION
**Pull request by bwbeach**
Not ready to merge yet, but would like comments.

### CODE REVIEW

File: b2/b2http.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+    except requests.ConnectionError as e0:
+        e1 = e0.args[0]
+        if isinstance(e1, requests.packages.urllib3.exceptions.MaxRetryError):
+            msg = e1.args[0]
+            if 'nodename nor servname provided, or not known' in msg:
+                # Unknown host, or DNS failing.  In the context of calling
+                # B2, this means that something is down between here and
+                # Backblaze, so we treat it like 503 Service Unavailable.
+                raise WebApiError(503, 'unknown_host', 'unable to locate host')
```
Comment (ppolewicz) [inline b2/b2http.py]: This is misleading: 503 means good connection to load balancer/front server,  but backend server error/all backends broken or busy

File: b2/b2http.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        elif isinstance(e1, requests.packages.urllib3.exceptions.ProtocolError):
+            e2 = e1.args[1]
+            if isinstance(e2, socket.error):
+                if e2.args[1] == 'Broken pipe':
+                    # Broken pipes are usually caused by the service rejecting
+                    # an upload request for cause, so we use a 400 Bad Request
```
Comment (ppolewicz) [inline b2/b2http.py]: Comment inconsistent with the code

File: b2/b2http.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                    # code.
+                    raise WebApiError(503, 'broken_pipe', 'unable to send entire request')
+        raise WebApiError(503, 'connection_error', 'unable to connect')
+
+    except Exception as e:
+        # Don't expect this to happen.  To get lots of info for
+        # debugging, call print_exception.
+        # print_exception(e)
```
Comment (ppolewicz) [inline b2/b2http.py]: Commented code

File: b2/b2http.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+def post(url, headers, data):
+    """
+    Posts to the given URL, translating any errors into a raised WebApiError.
+
+    :param url: The URL to post to.
+    :param headers: Headers to send.
+    :param data: Data to send: should be bytes.
+    :return: A requests object with the response.
+    """
+    return _translate_errors(lambda: requests.post(url, headers=headers, data=data))
```
Comment (ppolewicz) [inline b2/b2http.py]: `functools.partial` instead of lambda

File: b2/b2http.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    # Generic connection error
+    print('TEST: generic connection error')
+    try:
+        post('https://api.backblaze.com:6666/bad_port', {}, six.b(chr(0)) * 10000000)
+        assert False, 'should have failed with broken pipe'
+    except WebApiError as e:
+        assert e == WebApiError(503, 'connection_error', 'unable to connect')
+
+
+_test()
```
Comment (ppolewicz) [inline b2/b2http.py]: Should not be called here

File: b2/b2http.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        super(WebApiError, self).__init__(status, code, message)
+        self.status = status
+        self.code = code
+        self.message = message
+
+    def __repr__(self):
+        return "WebApiError(%d, '%s', '%s')" % (self.status, self.code, self.message)
+
+    def __eq__(self, other):
+        return repr(self) == repr(other)
```
Comment (ppolewicz) [inline b2/b2http.py]: This seems weird, exceptions don't usually have this. It is needed for tests, but checking `code` field is enough there. In test we want to check behavior, not identity.

Comment (bwbeach) [inline b2/b2http.py]: I'll go with `code`.

File: b2/b2http.py (hunk 4 above)
Comment (bwbeach) [inline b2/b2http.py]: Why?  I find the lambda here concise and clear.

File: b2/b2http.py (hunk 1 above)
Comment (bwbeach) [inline b2/b2http.py]: 503 can also be returned by a proxy in between you and the server.  If you think of this code as a proxy, it makes sense.

File: b2/b2http.py (hunk 4 above)
Comment (bwbeach) [inline b2/b2http.py]: Found a [StackOverflow article](https://stackoverflow.com/questions/3252228/python-why-is-functools-partial-necessary).  I'm willing to follow Guido on this one.

(And yes, I did do a lot of Lisp programming early in my career.)

File: b2/b2http.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    The operations supported are:
+       - post_json_return_json
+       - post_content_return_json
+       - get_content
+
+    The methods that return JSON either return a Python dict  or
+    raise a subclass of B2Error.  They can be used like this:
+
+        try:
+            response_dict = b2_http.post_json_return_json(url, headers, params)
```
Comment (ppolewicz) [inline b2/b2http.py]: This docstring should be for method, not for the class

File: b2/b2http.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        response = _translate_errors(
+            functools.partial(
+                requests.post,
+                url,
+                headers=headers,
+                data=data
+            )
+        )
+        try:
+            return json.loads(response.content.decode('utf-8'))
```
Comment (ppolewicz) [inline b2/b2http.py]: Is empty or non-json response handled?

File: b2/raw_api.py [hunk 9]
```diff
@@ -223,7 +123,8 @@ def _post_json(self, base_url, api_name, auth, **params):
         :return:
         """
         url = base_url + '/b2api/v1/' + api_name
-        return post_json(url, params, auth)
+        headers = dict(Authorization=auth)
```
Comment (ppolewicz) [inline b2/raw_api.py]: We usually use `{'Authorization': auth}` syntax

File: b2/b2http.py (hunk 7 above)
Comment (bwbeach) [inline b2/b2http.py]: OK.  It will duplicate some information, but you're right.

File: b2/b2http.py (hunk 8 above)
Comment (bwbeach) [inline b2/b2http.py]: No.  I'll add a comment.

**Pull request merged**